### ISSUE TITLE
Could you explain the training option "focal" ?

### ISSUE THREAD

**Issue by GabbySuwichaya** [zju3dv/LoFTR#43]
Hi ! Thank you very much for releasing the implementation.

As I am learning about this technology, I would like to kindly ask about the training option for  `self.loss_config['coarse_type'] == 'focal' ` ? 

What are the alpha and gamma for ? 
Specifically, why did you shape the supervised loss as follows: 
```
loss_pos = - alpha * torch.pow(1 - conf[pos_mask], gamma) * (conf[pos_mask]).log()
loss_neg = - alpha * torch.pow(conf[neg_mask], gamma) * (1 - conf[neg_mask]).log() 
```
And how are the alpha and gamma related to the previous supervised loss in 'cross_entropy' option?  
```
loss_pos  = - torch.log(conf[pos_mask]) 
 loss_neg = - torch.log(1 - conf[neg_mask]) 
```

And what are `torch.pow(1 - conf[pos_mask], gamma) ` and `torch.pow(conf[neg_mask], gamma) ` for ?

**Comment by zehongs** [comment]
Hi, sorry for the late reply!
The 'focal' represents focal loss, from [Focal Loss for Dense Object Detection](https://arxiv.org/pdf/1708.02002.pdf).  You can find more information about the alpha and gamma there.
Generally speaking, it gives better supervision when the positive and negative labels are unbalanced. In our case, there're more negative examples than the positive ones.

**Comment by GabbySuwichaya** [comment]
Thanks. 
Also, I have a question regarding your remark on the [Update strategy] https://github.com/zju3dv/LoFTR/blob/master/docs/TRAINING.md#updated-training-strategy).

> In the released training code, we use a slightly modified version of the coarse-level training supervision comparing to the one described in our paper. For example, as described in our paper, we only supervise the ground-truth positive matches when training the dual-softmax model. However, the entire confidence matrix produced by the dual-softmax matcher is supervised by default in the released code, regardless of the use of softmax operators. This implementation is counter-intuitive and unusual but leads to better evaluation results on estimating relative camera poses.  

What do you mean by `the entire confidence matrix produced by the dual-softmax matcher is supervised by default in the released code,  regardless of the use of softmax operators` ? 

It seems that the dual-softmax has the softmax operation [Line 119 (coarse_matching)](https://github.com/zju3dv/LoFTR/blob/261baf641cb9ada07dd9746e420ada7fe8a03152/src/loftr/utils/coarse_matching.py#L119) 

But what do you mean by `regardless of the use of softmax operators` ? 

```            
conf_matrix = F.softmax(sim_matrix, 1) * F.softmax(sim_matrix, 2)
``` 
Which softmax operation you are referring to?

**Comment by zehongs** [comment]
Oh, intuitively, we can compute loss w.r.t. the positive label only if using SoftMax(or dual-softmax), right?
And `the entire confidence matrix produced by the dual-softmax matcher is supervised by default in the released code`, this is to say that we also compute loss on the negative labels.

**Comment by GabbySuwichaya** [comment]
I see. Do you mean that the entire confidence matrix is the confidence for both positive and negative labels... (whether or not there is a soft-max operation after it)?

**Comment by zehongs** [comment]
Yes.

**Comment by GabbySuwichaya** [comment]
Thank you very much. I understand the work more now. :)
